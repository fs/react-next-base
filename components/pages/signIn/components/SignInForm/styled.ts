import styled from 'styled-components';

export const FormContentWrapper = styled.div`
  width: 40rem;
  max-width: 100%;
`;

export const FieldWrapper = styled.div`
  margin-top: 1rem;
`;

export const SubmitButtonWrapper = styled.div`
  margin-top: 2rem;
`;
