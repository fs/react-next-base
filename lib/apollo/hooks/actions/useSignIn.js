import { useMutation } from '@apollo/client';
import { useRouter } from 'next/router';
import { HOME } from 'config/routes';
import { SIGN_IN_EVENT } from 'config/globalEvents.json';

import SignIn from 'graphql/mutations/signIn.graphql';
import CurrentUser from 'graphql/queries/currentUser.graphql';

import useNotifier from 'hooks/useNotifier';

const useSignIn = () => {
  const { setError } = useNotifier();
  const router = useRouter();

  const [mutation, mutationState] = useMutation(SignIn, {
    update: (store, { data }) => {
      store.writeQuery({
        query: CurrentUser,
        data: {
          me: {
            ...data.signin.me,
          },
        },
      });
    },
  });

  const mutate = async ({ email, password }) => {
    const signInInput = {
      email,
      password,
    };

    try {
      await mutation({ variables: { input: signInInput } });

      window.localStorage.setItem(SIGN_IN_EVENT, Date.now());

      router.push(HOME);
    } catch (error) {
      setError(error);
    }
  };

  return [mutate, mutationState];
};

export default useSignIn;
