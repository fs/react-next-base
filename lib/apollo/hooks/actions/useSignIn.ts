import { useRouter } from 'next/router';

import { HOME } from 'config/routes';
import { SIGN_IN_EVENT } from 'config/globalEvents.json';
import { useNotifier } from 'contexts/NotifierContext';

import type { SignInVariables } from 'api/types/user/signInApiType';
import type { SignInMutationResult } from 'api/mutations/useSignInMutation';
import useSignInMutation from 'api/mutations/useSignInMutation';

const useSignIn = (): [(variables: SignInVariables) => Promise<void>, SignInMutationResult] => {
  const { setError } = useNotifier();
  const router = useRouter();

  const [mutation, mutationResult] = useSignInMutation();

  const mutate = async ({ email, password }: SignInVariables) => {
    const signInInput = {
      email,
      password,
    };

    try {
      await mutation({ variables: { input: signInInput } });

      window.localStorage.setItem(SIGN_IN_EVENT, Date.now().toString());

      router.push(HOME);
    } catch (error) {
      if (setError) setError(error);
    }
  };

  return [mutate, mutationResult];
};

export default useSignIn;
