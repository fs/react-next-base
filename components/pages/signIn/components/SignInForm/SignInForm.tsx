import React from 'react';
import { Form, Formik, FormikProps } from 'formik';
import * as Yup from 'yup';

import useSignIn from 'lib/apollo/hooks/actions/useSignIn';

import Button from 'components/shared/atoms/Button';
import FormFieldInput from 'components/shared/atoms/FormField';
import Loader from 'components/shared/atoms/Loader';
import { FormFieldType } from 'types/formsType';

import { FieldWrapper, FormContentWrapper, SubmitButtonWrapper } from './styled';

const initialValues = {
  email: '',
  password: '',
};

const SignInValidationSchema = Yup.object().shape({
  email: Yup.string().email('Must be a valid email').max(255).required('This field is required'),
  password: Yup.string().required('This field is required'),
});

type ValuesFromFormik = Parameters<ReturnType<typeof useSignIn>[0]>[0];

const SignInFormContent = ({ isSubmitting }: FormikProps<ValuesFromFormik>) => (
  <FormContentWrapper>
    <Form>
      <FieldWrapper>
        <FormFieldInput name="email" type={FormFieldType.email} label="Email" />
      </FieldWrapper>
      <FieldWrapper>
        <FormFieldInput name="password" type={FormFieldType.password} label="Password" />
      </FieldWrapper>
      <SubmitButtonWrapper>
        <Button type={FormFieldType.submit} testID="submit-button" disabled={isSubmitting}>
          Submit
        </Button>
      </SubmitButtonWrapper>
    </Form>
  </FormContentWrapper>
);

const SignInForm = () => {
  const [signIn, signInState] = useSignIn();

  const onSubmit = async (values: ValuesFromFormik) => {
    try {
      await signIn(values);
    } catch (error) {
      console.error(error);
    }
  };

  return (
    <>
      <Formik
        initialValues={initialValues}
        onSubmit={onSubmit}
        component={SignInFormContent}
        validationSchema={SignInValidationSchema}
      />
      {signInState.loading && <Loader testId="signin-loader">Loading...</Loader>}
    </>
  );
};

export default SignInForm;
