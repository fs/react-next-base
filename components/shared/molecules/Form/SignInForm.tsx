import React from 'react';
import styled from 'styled-components';
import { Form, Formik, FormikHelpers, FormikProps } from 'formik';
import * as Yup from 'yup';

import useSignIn from 'lib/apollo/hooks/actions/useSignIn';
import { SubmitButton } from 'components/shared/molecules/Form/formFields';
import { FormFieldType } from './forms.types';
import FormField from '../../atoms/FormField';

const FormContentWrapper = styled.div`
  max-width: 40rem;
`;

const initialValues = {
  email: '',
  password: '',
};

const SignInValidationSchema = Yup.object().shape({
  email: Yup.string().required('Required'),
  password: Yup.string().required('Required'),
});

type ValuesFromFormik = Parameters<ReturnType<typeof useSignIn>[0]>[0];

const SignInFormContent = ({ isSubmitting }: FormikProps<ValuesFromFormik>) => (
  <FormContentWrapper>
    <Form>
      <FormField name="email" type={FormFieldType.email} label="Email" />
      <FormField name="password" type={FormFieldType.password} label="Password" />
      <SubmitButton type={FormFieldType.submit} name="signIn" isFormSubmitting={isSubmitting}>
        Submit
      </SubmitButton>
    </Form>
  </FormContentWrapper>
);

const SignInForm = () => {
  const [signIn] = useSignIn();

  const onSubmit = async (values: ValuesFromFormik, { setSubmitting }: FormikHelpers<ValuesFromFormik>) => {
    try {
      setSubmitting(true);
      await signIn(values);
      setSubmitting(false);
    } catch (error) {
      console.error(error);
    }
  };

  return (
    <Formik
      initialValues={initialValues}
      onSubmit={onSubmit}
      component={SignInFormContent}
      validationSchema={SignInValidationSchema}
    />
  );
};

export default SignInForm;
