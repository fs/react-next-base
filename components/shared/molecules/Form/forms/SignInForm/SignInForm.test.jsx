import React from 'react';
import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import 'jest-styled-components';
import renderWithTheme from '__tests__/helpers/renderWithTheme';

import useSignIn from 'lib/apollo/hooks/actions/useSignIn';
import SignInForm from './SignInForm';

jest.mock('lib/apollo/hooks/actions/useSignIn');

describe('SignInForm', () => {
  let signIn;
  let signInState;

  beforeEach(() => {
    signIn = jest.fn();
    signInState = {
      error: undefined,
      loading: null,
      data: {
        signIn: {},
      },
    };
    useSignIn.mockImplementation(() => [signIn, signInState]);
  });

  test('should render correctly', () => {
    // Act
    const { container } = render(renderWithTheme(<SignInForm />));

    // Assert
    expect(container).toMatchSnapshot();
  });

  test('should call SignIn fn after click "Submit" button', async () => {
    // Arrange
    const inputEmailValue = 'email@gmail.com';
    const inputPasswordValue = 'password';

    // Act
    render(renderWithTheme(<SignInForm />));

    const inputEmail = screen.getByTestId('input-email');
    const inputPassword = screen.getByTestId('input-password');

    fireEvent.change(inputEmail, { target: { value: inputEmailValue } });
    fireEvent.change(inputPassword, { target: { value: inputPasswordValue } });

    expect(inputEmail.value).toBe(inputEmailValue);
    expect(inputPassword.value).toBe(inputPasswordValue);

    fireEvent.click(screen.getByTestId('submit-button'));

    await waitFor(() => {
      expect(signIn).toHaveBeenCalled();
    });
  });

  test('should show error if fields are filled incorrectly', async () => {
    // Arrange
    const invalidInputEmail = 'test';
    const invalidInputPassword = '';
    render(renderWithTheme(<SignInForm />));
    const inputEmail = screen.getByTestId('input-email');
    const inputPassword = screen.getByTestId('input-password');
    const submitButton = screen.getByTestId('submit-button');

    fireEvent.change(inputEmail, { target: { value: invalidInputEmail } });
    fireEvent.click(submitButton);
    await waitFor(() => {
      expect(screen.queryByText('Must be a valid email')).toBeInTheDocument();
    });

    fireEvent.change(inputEmail, { target: { value: '' } });
    await waitFor(() => {
      expect(screen.queryByText('This field is required')).toBeInTheDocument();
    });

    fireEvent.change(inputPassword, { target: { value: invalidInputPassword } });
    fireEvent.click(submitButton);
    await waitFor(() => {
      expect(screen.queryAllByText('This field is required')).toHaveLength(2);
    });
  });

  test('should show loader if is loading', () => {
    // Arrange
    signInState = {
      error: undefined,
      loading: true,
      data: null,
    };

    useSignIn.mockImplementation(() => [signIn, signInState]);
    render(renderWithTheme(<SignInForm />));

    const inputEmailValue = 'email@gmail.com';
    const inputPasswordValue = 'password';

    const inputEmail = screen.getByTestId('input-email');
    const inputPassword = screen.getByTestId('input-password');

    // Act
    fireEvent.change(inputEmail, { target: { value: inputEmailValue } });
    fireEvent.change(inputPassword, { target: { value: inputPasswordValue } });
    fireEvent.click(screen.getByTestId('submit-button'));

    // Assert
    expect(screen.getByTestId('signin-loader')).toBeInTheDocument();
  });
});
