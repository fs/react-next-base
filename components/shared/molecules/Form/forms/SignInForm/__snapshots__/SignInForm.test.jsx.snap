// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignInForm should render correctly 1`] = `
.c5 {
  background-color: #eee;
  height: 3rem;
  width: 100%;
  font-size: 1rem;
  border-radius: 0.3rem;
  border: 1px solid #ccc;
}

.c5:hover {
  background-color: #ccc;
  border: 1px solid #606c76;
}

.c2 {
  width: 100%;
  padding: 1rem;
  font-size: 1rem;
  border-radius: 0.3rem;
  border: 1px solid #ccc;
}

.c3 {
  margin-top: 0.5rem;
}

.c0 {
  width: 40rem;
}

.c1 {
  margin-top: 1rem;
}

.c4 {
  margin-top: 2rem;
}

<div>
  <div
    class="c0"
  >
    <form
      action="#"
    >
      <div
        class="c1"
      >
        <label
          for="email"
        >
          Email
          <input
            autocomplete="on"
            class="c2 c3"
            data-testid="input-email"
            id="email"
            name="email"
            type="email"
            value=""
          />
        </label>
      </div>
      <div
        class="c1"
      >
        <label
          for="password"
        >
          Password
          <input
            autocomplete="on"
            class="c2 c3"
            data-testid="input-password"
            id="password"
            name="password"
            type="password"
            value=""
          />
        </label>
      </div>
      <div
        class="c4"
      >
        <button
          class="c5"
          data-cy="submit-button"
          data-testid="submit-button"
          type="submit"
        >
          Submit
        </button>
      </div>
    </form>
  </div>
</div>
`;
